### PR TITLE
Skip INSFORGE_INTERNAL_URL cloud seed

### PR DESCRIPTION
**Pull request by malsomesh9**
## Summary
- stop seeding `INSFORGE_INTERNAL_URL` in cloud environments
- keep rewriting `INSFORGE_INTERNAL_URL` to `INSFORGE_BASE_URL` only in cloud so OSS keeps its internal container route
- add focused regression tests covering both source locations

## Why
Cloud deployments never use the hardcoded internal container URL, but the reserved secret still shows up and confuses function authors about whether they should use `INSFORGE_INTERNAL_URL` or `INSFORGE_BASE_URL`.

## Validation
- `cd backend && npm test -- --run tests/unit/seed-jwt-secret.test.ts tests/unit/function-secrets-cloud-behavior.test.ts`

---
## Summary by cubic
Stops seeding `INSFORGE_INTERNAL_URL` in cloud and gates the `INSFORGE_INTERNAL_URL` → `INSFORGE_BASE_URL` rewrite behind a cloud check, reducing confusion while keeping OSS internal routing intact.

- **Bug Fixes**
  - Skip creating the `INSFORGE_INTERNAL_URL` secret during cloud seeding.
  - Rewrite `INSFORGE_INTERNAL_URL` to `INSFORGE_BASE_URL` only when `isCloudEnvironment()` is true; OSS keeps container-to-container routing.
  - Added runtime unit tests for cloud/OSS secret behavior (rewrite and seeding), and ensured `JWT_SECRET` still seeds in cloud.

<sup>Written for commit f6cefa2ce1aac2f341babe6531747f8881668cec. Summary will update on new commits. <a href="https://cubic.dev/pr/InsForge/InsForge/pull/1298?utm_source=github">Review in cubic</a></sup>

> [!NOTE]
> ### Skip seeding `INSFORGE_INTERNAL_URL` secret in cloud environments
> - `seedBackend` in [seed.ts](https://github.com/InsForge/InsForge/pull/1298/files#diff-63196e4b85e26873baaf8f4688c0de418f99bafec3af5143fdae6a9d44f6e9b7) no longer creates the `INSFORGE_INTERNAL_URL` reserved secret when running in a cloud environment.
> - `FunctionService.getFunctionSecrets` in [function.service.ts](https://github.com/InsForge/InsForge/pull/1298/files#diff-574d7a1f366ea56db7647fc0040c3b519eb81aa6cb3b3f2db71839a46cfaecbc) now only rewrites `INSFORGE_INTERNAL_URL` to `INSFORGE_BASE_URL` in cloud environments, preserving the original value in OSS environments.
> - Unit tests added to verify both behaviors across cloud and OSS environments.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized f6cefa2.</sup>
> 
>

## Summary by CodeRabbit

* **Bug Fixes**
  * Secret handling now differentiates cloud vs non-cloud: internal URL is rewritten in cloud deployments and preserved for on-prem/OSS.
  * Reserved internal URL seeding only runs in non-cloud environments.

* **Tests**
  * Added tests validating cloud vs non-cloud secret behavior.
  * Rewrote seed initialization tests to assert environment-specific seeding outcomes.

[Review Change Stack](https://app.coderabbit.ai/change-stack/InsForge/InsForge/pull/1298?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: backend/tests/unit/function-secrets-cloud-behavior.test.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+describe('FunctionService cloud secret behavior', () => {
+  const currentDir = path.dirname(fileURLToPath(import.meta.url));
+  const functionServicePath = path.resolve(currentDir, '../../src/services/functions/function.service.ts');
+  const functionServiceSource = fs.readFileSync(functionServicePath, 'utf8');
+
+  it('only rewrites INSFORGE_INTERNAL_URL in cloud environments', () => {
+    expect(functionServiceSource).toContain('isCloudEnvironment() && baseUrlValue');
+    expect(functionServiceSource).toContain("secretMap['INSFORGE_INTERNAL_URL'] = baseUrlValue");
+  });
```
[the remark this answers is not in the document]
Comment (malsomesh9) [reply backend/tests/unit/function-secrets-cloud-behavior.test.ts]: Updated in f6cefa2c. I replaced the source-text assertion with runtime tests that mock `isCloudEnvironment` and the secret service, then assert the returned `secretMap` behavior directly for both cloud and OSS paths.

File: backend/tests/unit/function-secrets-cloud-behavior.test.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+});
```
[the remark this answers is not in the document]
Comment (malsomesh9) [reply backend/tests/unit/function-secrets-cloud-behavior.test.ts]: Fixed by the same update in f6cefa2c. The old file was replaced, and `npx prettier --check tests/unit/seed-jwt-secret.test.ts tests/unit/function-secrets-cloud-behavior.test.ts` now passes locally.

File: backend/tests/unit/function-secrets-cloud-behavior.test.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { describe, expect, it } from 'vitest';
```
[the remark this answers is not in the document]
Comment (malsomesh9) [reply backend/tests/unit/function-secrets-cloud-behavior.test.ts]: Addressed in f6cefa2c. This test now exercises runtime behavior by mocking the environment and secret lookups instead of matching source strings.

File: backend/tests/unit/seed-jwt-secret.test.ts [hunk 4]
```diff
@@ -21,4 +21,10 @@ describe('seedBackend JWT_SECRET initialization', () => {
     expect(seedSource).toContain("getSecretByKey('JWT_SECRET')");
     expect(seedSource).toMatch(/existingJwtSecret\s*===\s*null/);
   });
+
+  it('only seeds INSFORGE_INTERNAL_URL outside cloud environments', () => {
+    expect(seedSource).toMatch(
+      /if\s*\(!isCloudEnvironment\(\)\)\s*\{[\s\S]*getSecretByKey\('INSFORGE_INTERNAL_URL'\)/
```
[the remark this answers is not in the document]
Comment (malsomesh9) [reply backend/tests/unit/seed-jwt-secret.test.ts]: Addressed in f6cefa2c. The seed test now runs `seedBackend()` with mocked collaborators and verifies cloud vs OSS behavior directly, including the `JWT_SECRET` path.

File: backend/src/utils/seed.ts [hunk 5]
```diff
@@ -276,17 +276,19 @@ export async function seedBackend(): Promise<void> {
...
-      logger.info('✅ INSFORGE_INTERNAL_URL secret initialized');
+      if (existingInternalUrlSecret === null) {
+        await secretService.createSecret({
+          key: 'INSFORGE_INTERNAL_URL',
+          isReserved: true,
+          value: insforgInternalUrl,
+        });
+        logger.info('✅ INSFORGE_INTERNAL_URL secret initialized');
+      }
     }
```
[the remark this answers is not in the document]
Comment (malsomesh9) [reply backend/src/utils/seed.ts:292]: Good call on the migration angle. I kept this PR scoped to the behavior change for new cloud deployments plus the cloud-only rewrite for existing deployments at function inject time; DB cleanup for already-seeded cloud projects would be a broader migration/follow-up change rather than a small fix in this PR.

**Comment by malsomesh9** [comment]
@tonychang04 what are your views on it??

**Review by tonychang04 — APPROVED** [review]

**Review by jwfing — COMMENTED** [review]
## Summary

Correctly gates `INSFORGE_INTERNAL_URL` seeding and the `→ INSFORGE_BASE_URL` rewrite behind `isCloudEnvironment()`, with runtime unit tests replacing the previous source-inspection anti-pattern.

## Requirements context

No spec/plan under `/docs/superpowers/specs/` or `/docs/superpowers/plans/` matches this change — review is assessed against the PR description and code alone.

---

## Findings

### Critical

*(none)*

---

### Suggestion

**[Software Engineering] `function-secrets-cloud-behavior.test.ts` tests a private method via type-casting**

`backend/tests/unit/function-secrets-cloud-behavior.test.ts:53-72`

```ts
const service = FunctionService.getInstance() as unknown as {
  getFunctionSecrets: () => Promise<Record<string, string>>;
};
```

`getFunctionSecrets` is `private`. Testing it via `as unknown as` bypasses TypeScript visibility and couples the test to the internal name — if the method is renamed or inlined the tests will silently stop covering the behaviour. Consider testing through the public method that calls `getFunctionSecrets` (e.g. `deployFunction` or `runFunction`, whichever surfaces the secret map), so the test contract stays stable against internal refactors.

---

**[Functionality] Missing edge-case: cloud environment where `INSFORGE_BASE_URL` is not yet seeded**

`backend/src/services/functions/function.service.ts:821`

```ts
if (isCloudEnvironment() && baseUrlValue && secretMap['INSFORGE_INTERNAL_URL']) {
```

If a cloud instance was previously seeded with `INSFORGE_INTERNAL_URL` (pre-fix deployment) but for some reason `INSFORGE_BASE_URL` doesn't exist (`baseUrlValue` is `null`), the rewrite silently no-ops and functions receive the container URL (`http://insforge:7130`), which is unreachable from Deno Subhosting. This edge case is unlikely in normal operation but could be confusing to debug. A `logger.warn` when `isCloudEnvironment() && secretMap['INSFORGE_INTERNAL_URL'] && !baseUrlValue` would make the failure visible. No test covers this path.

---

**[Software Engineering] `seed-jwt-secret.test.ts` OSS test: `ANON_KEY` seeding side-effect not asserted**

`backend/tests/unit/seed-jwt-secret.test.ts:100-114`

The OSS test calls the real `seedBackend()` which will also create `ANON_KEY`, `INSFORGE_BASE_URL`, etc. (all mock `getSecretByKey` calls return `null`). The test only asserts on `INSFORGE_INTERNAL_URL`. This is fine as written, but a future maintainer adding a new secret to `seedBackend()` won't get a test failure if they forget the cloud guard — consider adding a `toHaveBeenCalledTimes` assertion or at least a comment noting the intent.

---

### Information

**[Software Engineering] Pre-existing typo: `insforgInternalUrl`** (`seed.ts:281`) — missing `e` in `insforge`. Not introduced by this PR; mentioning it for the record.

**[Software Engineering] Positive note:** Replacing the old `fs.readFileSync` source-inspection tests with actual runtime execution + mocks is a meaningful improvement in test quality. The new `seed-jwt-secret.test.ts` tests will catch real regressions that the string-scanning approach would miss.

---

## Verdict

**approved** *(informational — human must give the explicit GitHub approval)*

No blocking issues. The implementation is correct: new cloud deployments won't see a spurious `INSFORGE_INTERNAL_URL`, existing cloud deployments with the secret already seeded are handled by the rewrite guard, and OSS container routing is preserved. The suggestions above are worth a follow-up but don't block merging.

**Comment by tonychang04** [comment]
I believe the code looks good and can go to production!
